File: robotarm_control_cpp/src/unityBluetooth.cpp
#include <chrono>
#include <cstdlib>
#include <memory>
#include "color_srv/srv/detection.hpp"
#include "rclcpp/rclcpp.hpp"

using Detection = color_srv::srv::Detection;

int main(int argc, char * argv[])
{
	rclcpp::init(argc, argv);
	auto node = rclcpp::Node::make_shared("picking_client");
	auto client = node->create_client<Detection>("picking");
	auto request = std::make_shared<Detection::Request>();
	request->color = "red";
	request->wrist = 10;
	request->wristv = 20;
	request->elbow = 30;
	request->shoulder = 40;
	request->pshoulder = 50;
	while (!client->wait_for_service(std::chrono::seconds(1)))
	{
		if(!rclcpp::ok()){
			RCLCPP_ERROR(node->get_logger(), "client interrupted while waiting for service to appear.");
			return 1;
		}
		RCLCPP_INFO(node->get_logger(), "waiting for service to appear");
	}
	auto result_future = client->async_send_request(request);
	if(rclcpp::spin_until_future_complete(node, result_future) != rclcpp::executor::FutureReturnCode::SUCCESS)
	{
		RCLCPP_ERROR(node->get_logger(), "service call failed");
		return 1;
	}
	auto result = result_future.get();
	RCLCPP_INFO(node->get_logger(), "result %d", result->result);
	rclcpp::shutdown();
	return 0;
}

File: ros2_robotArm_control/ros2_robotarm_control/unityBluetooth.py
#!/usr/bin/env python

import rclpy
from rclpy.node import Node
from color_srv.srv import Detection
#import cv2
import numpy as np
from bluetooth import *
import time

class Unity_bluetooth(Node):
	def __init__(self):
		super().__init__('picking_client')
		self.clnt = self.create_client(Detection, 'picking')
		while not self.clnt.wait_for_service(timeout_sec=1.0):
			self.get_logger().info('service not available')
		self.req = Detection.Request()
		self.data = ""
		self.data_split = []
		
	def bluetooth_conn(self):
		print("bluetooth_conn()")
		server_socket = BluetoothSocket(RFCOMM)
		server_socket.bind(('', 1))
		server_socket.listen(1)

		client_socket, address = server_socket.accept()
		
		self.data = client_socket.recv(1024).decode('ascii')
		self.data_split = self.data.split('/')
		self.req.color = self.data_split[0]
		self.req.pshoulder = int(self.data_split[1])
		self.req.shoulder = int(self.data_split[2])
		self.req.elbow = int(self.data_split[3])
		self.req.wristv = int(self.data_split[4])
		self.req.wrist = int(self.data_split[5])

		self.future = self.clnt.call_async(self.req)
		print("bluetooth_connf")
		time.sleep(2)
		client_socket.close()
		server_socket.close()

		

def main(args=None):

	rclpy.init(args=args)

	unity_bluetooth = Unity_bluetooth()
	unity_bluetooth.bluetooth_conn()
	

	while rclpy.ok():
		rclpy.spin_once(unity_bluetooth)
		if unity_bluetooth.future.done():
			try:
				response = unity_bluetooth.future.result()
			except Exception as e:
				unity_bluetooth.get_logger().info('service fail')
		else:
			unity_bluetooth.get_logger().info('color: %s / x: %d / y: %d' % (unity_bluetooth.req.color, unity_bluetooth.req.wrist, unity_bluetooth.req.shoulder))
		break
	unity_bluetooth.destroy_node()
	rclpy.shutdown()

if __name__ == '__main__':
	main()
